libafl/src/executors/forkserver.rs

```diff
@@ -277,6 +263,68 @@ impl ConfigTarget for Command {
         // This calls our non-shady function from above.
         unsafe { self.pre_exec(func) }
     }
+
+    fn setdup2(&mut self, old_fd: RawFd, new_fd: RawFd) -> &mut Self {
+        let func = move || {
+            let ret = unsafe { libc::dup2(old_fd, new_fd) };
+            if ret < 0 {
+                return Err(io::Error::last_os_error());
+            }
+            Ok(())
+        };
+        // # Safety
+        // This calls our non-shady function from above.
+        unsafe { self.pre_exec(func) }
+    }
+}
+
+/// Thin wrapper of memfd
+#[derive(Debug)]
+struct MemFd {
```

Comment:
This should probably be in libafl_bolts

Reply:
Also, why not use the "normal" ShMem crate here?

Reply:
> Also, why not use the "normal" ShMem crate here?

We can not use `ShMem` i.e. `shm_open` because the fd returned is not writable on macOS and using `memfd_create` is straightforward and easier.

Reply:
Addon: Not using `MemfdShMem` because I want an `OwnedFd`.

Reply:
Should still go to bolts I think

Reply:
> > Also, why not use the "normal" ShMem crate here?
> 
> We can not use `ShMem` i.e. `shm_open` because the fd returned is not writable on macOS and using `memfd_create` is straightforward and easier.

Could we still use an ShMemProvider, and just add an extra function to get a non-writable map?
In a library it's not about making one particular use case easier, but more about making all use cases align with each other

Reply:
You didn’t get it. It is https://stackoverflow.com/questions/73752631/cant-write-to-fd-from-shm-open-on-macos so that shm_open is totally not working on macOS

Reply:
On Linux, ShmMem is working but there are teo reasons not using it:

1. It has an extra useless mmap operation
2. An OwnedFd is needed here instead of [u8; 20] so that we can have Rust to track the semantics.

Reply:
I'm not talking about shm_open, I'm talking about creating an MemFdShMemProvider and then adding a method to get the OwnedFd from there (or similar)

Reply:
Oh okay, I see. That makes much sense, will do.

Reply:
> An OwnedFd is needed here instead of [u8; 20] so that we can have Rust to track the semantics.

You can get to an OwnedFd from `u8; 20`, this is the way (Fds have numbers, so..)

Reply:
Yeah, if we have a provider we can take self then so that not breaking semantics. I just misunderstood that you wanna reuse current provider.

This reminds me that we should indeed have such providers.

Reply:
https://github.com/AFLplusplus/LibAFL/blob/ecaa013263f233065431adff35727f49a0417ac3/libafl_bolts/src/shmem.rs#L1412

Oh yeah it’s there and we need a new method like to_file(self) -> File

Reply:
And make mmap optional too